### ISSUE TITLE
Overriding python wrapper paths is painful/impossible

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#23639]
Context: I'd like to be able to have an efficient `buildbot.withPackages` because its configuration can request arbitrary python packages at runtime depending on the configuration. The derivation currently has a `plugins` parameter but that rebuilds the entire buildbot derivation (and runs the tests) each time, even though it doesn't really make any difference to the buildbot derivation and only affects the path we throw into the wrapper binary.

I poked around and found that wrapping python derivations isn't optional, in [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix#L79-L85). Furthermore, there are two levels of path injection: there's a python script that uses `functools.reduce` to inject the python library dependency paths into the search path in a python shim, and then the result of that still gets wrapped with `wrapProgram`.

Basically, I'm looking for a good way to defer all the magic python wrapping until later, when I know the full set of paths I want to inject. I can't really use the `python.withPackages` approach because that relies on how the python derivation works.

cc @FRidh @nand0p @domenkozar @garbas

**Comment by FRidh** [comment]
> I poked around and found that wrapping python derivations isn't optional, in here. Furthermore, there are two levels of path injection: there's a python script that uses functools.reduce to inject the python library dependency paths into the search path in a python shim, and then the result of that still gets wrapped with wrapProgram.

Those exist for Python applications. I'm not a fan of it, and would like to get rid of it. It's actually pretty simple; remove the wrapper scripts, and add a derivation that wraps the unwrapped package. That will allow you to use different packages during runtime.

I've been noticing this is a recurring issue, not just for Python in Nixpkgs, but also with other languages or frameworks in Nixpkgs. I think we need to think about a better, more generic way, to defer wrapping.

**Comment by copumpkin** [comment]
Yeah, probably makes sense. My main issue with undoing the wrapping myself is that I don't know how to replicate the default wrapping behavior myself, at a later time.

Part of it might be that `wrapPythonPrograms` is rather magical. I'll poke around and see if I can figure it out.

**Comment by copumpkin** [comment]
Never mind, I realized that `PYTHONPATH` does work and I was doing it wrong 😠
